### DISCUSSION TITLE
Hibernate panache same primary key for one to one relations

### DISCUSSION THREAD

**Discussion by haya-hammoud** [quarkusio/quarkus#25755] [Q&A] [answered]
Hi 
I am using Hibernate ORM with Panache to create my entities, anyone have an idea how to implement OneToOne association that assumes both entities share the same primary key values? 

My code looks like this 

```java
@Entity
public class Car extends PanacheEntity {

 public String plateNumber;
 public String color;

@OneToOne(optional = true, mappedBy = “car”, cascade = CascadeType.REMOVE, orphanRemoval = true)
    public Ownership ownership;
}

@Entity
public class Onwership extends PanacheEntity {

public String ownerName;

@OneToOne
@MapsId
@JoinColumn(name = "id")
	public Car car;

}
```

Now, when I’m trying to add ownership (persist) for an existing car, I got this 
org.hibernate.id.IdentifierGenerationException: attempted to assign id from null one-to-one property
I think there is sth wrong with my entities definition. I would appreciate any help.

[a comment by an automated account is not shown]

**Comment by yrodiere** [comment, marked as answer]
Hello,

I'm guessing either `Onwership#car` or `Car#ownership` is null?

It would help if you showed the actual code that persists the entities.

**Reply by yrodiere** [reply]
Ah, actually I didn't see you were using Panache. You'll probably want to extend `PanacheEntityBase` instead of `PanacheEntity` when you define your own identifier.

**Reply by haya-hammoud** [reply]
@yrodiere I can try using `PanacheEntityBase` but before that can I just ask if it's good practice to do this.
Following illustrations shows two entities that have a `OneToOne` relation.

![image](https://user-images.githubusercontent.com/22799651/170045703-9893f0b8-9169-41dc-9411-487b46594e11.png)

- Since a car have one owner and one owner can have only one car, I was thinking that both of these ID's marked in red could be the same. The benefit of doing it this way is that in the resource if I'm querying `/owner/{id}` or if I'm doing `/owner?car={id}` , both of the query would give the same result.
- My question is on best practice and if it's the right way to handle `OneToOne` relation as shown in the diagram - circle 3 colored blue.

My reference:
https://docs.oracle.com/javaee/6/api/javax/persistence/OneToOne.html

```java
 Example 2: One-to-one association that assumes both the source and target share the same primary key values. 

    // On Employee class:

    @Entity
    public class Employee {
        @Id Integer id;
    
        @OneToOne @MapsId
        EmployeeInfo info;
        ...
    }
```

**Reply by yrodiere** [reply]
Well that's a lot of trust to put in me, a random stranger on the internet, but okay :) Just keep in mind you can ask the same question to 10 different people and get 10 different answers.

Personally I wouldn't do this unless I have specific reasons, because this complicates your domain model and that can have consequences down the line. For example if you evolve your model and decide that `Ownership` becomes `Owner` and an owner can trade his car with another, you'll be in trouble.

One case where it could make sense is if the ID is actually something that has meaning to users, not just some generated integer. But even then you could just use a generated id plus another property annotated with `@NaturalId`.

Another reason to do it could be performance, maybe. Though I won't try to find an example.

If you neither have a business reason or a performance reason to do this, I'd say don't bother.

I believe `@MapsId` and similar annotations are mainly there for legacy applications, those that existed *before* they switched to JPA. If you're designing a new model from scratch... Keep it simple.

**Reply by haya-hammoud** [reply]
I agree, I think I will just keep it simple by just using a foreign key column
Thanks a lot for the explanation :)
